File: examples/python/src/docimprint_examples/production_flow.py
"""Evidence + receipt deep dive: URL extract → verify → optional notarize → receipt audit.

Usage:
  python -m docimprint_examples.production_flow --url https://example.com/contract.pdf
  python -m docimprint_examples.production_flow --url https://example.com/contract.pdf --notarize
"""

from __future__ import annotations

import argparse
from typing import Any

from docimprint import DocImprintError
from rich.panel import Panel

from docimprint_examples.shared import (
    api_verify_url,
    bundle_id_of,
    console,
    json_dump,
    load_config,
    manifest_sha256,
    open_session,
    print_error,
    print_share_panel,
    print_warnings,
    receipt_id,
    require_valid_receipt,
    verify_tool_url,
    write_evidence_pack,
)


def run(
    *,
    url: str,
    notarize: bool,
    api_key: str,
    base_url: str,
    allow_missing_receipt: bool = False,
) -> dict[str, Any]:
    summary: dict[str, Any] = {
        "demo": "production_flow",
        "source_url": url,
        "bundle_id": None,
        "manifest_sha256": None,
        "verify": None,
        "notarize": None,
        "inline_receipt_id": None,
        "receipt_list_count": 0,
        "receipt_verify": None,
        "warnings": [],
    }

    console.print(
        Panel.fit(
            "[bold]Evidence + receipt[/]\n"
            "Extract a URL into a signed bundle, verify integrity, "
            "then independently audit the action receipt.",
            border_style="blue",
        )
    )

    with open_session(api_key, base_url) as session:
        extracted = session.client.extract(url=url)
        bundle_id = bundle_id_of(extracted)
        if not bundle_id:
            raise DocImprintError(
                "extract response missing bundle_id",
                code="MISSING_BUNDLE_ID",
                details=extracted,
            )

        manifest = manifest_sha256(extracted)
        summary["bundle_id"] = bundle_id
        summary["manifest_sha256"] = manifest
        console.print(f"  bundle_id = [cyan]{bundle_id}[/]")
        console.print(f"  manifest  = [cyan]{manifest or '—'}[/]")

        verified = session.client.verify(bundle_id)
        inline_receipt = (
            verified.get("receipt") if isinstance(verified.get("receipt"), dict) else None
        )
        summary["verify"] = {
            "valid": bool(verified.get("valid")),
            "manifest_sha256": manifest_sha256(verified) or manifest,
            "checks": verified.get("checks"),
        }
        summary["inline_receipt_id"] = receipt_id(inline_receipt)

        if not verified.get("valid"):
            raise DocImprintError(
                "bundle verification failed",
                code="VERIFY_INVALID",
                details=verified,
            )
        console.print("  verify   = [green]valid[/]")

        if notarize:
            notarized = session.client.notarize(bundle_id)
            notarize_receipt = (
                notarized.get("receipt") if isinstance(notarized.get("receipt"), dict) else None
            )
            summary["notarize"] = {
                "attestation": notarized.get("attestation"),
                "receipt_id": receipt_id(notarize_receipt),
            }
            console.print("  notarize = done")

        receipt_verify, receipt_rows, receipt_warnings = require_valid_receipt(
            session.client,
            bundle_id=bundle_id,
            preferred_id=summary["inline_receipt_id"],
            allow_missing=allow_missing_receipt,
        )
        summary["warnings"].extend(receipt_warnings)
        summary["receipt_list_count"] = len(receipt_rows)
        summary["receipt_verify"] = receipt_verify
        if receipt_verify:
            console.print("  receipt  = [green]valid[/]")

        summary["links"] = {
            "verify_ui": verify_tool_url(bundle_id),
            "verify_api": api_verify_url(base_url, bundle_id),
        }
        pack_path = write_evidence_pack(bundle_id, summary)
        print_share_panel(
            bundle_id=bundle_id,
            manifest=summary["manifest_sha256"],
            base_url=base_url,
            receipt_verify=summary.get("receipt_verify"),
            pack_path=pack_path,
        )
        print_warnings(summary["warnings"])
        return summary


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description=(
            "Production DocImprint flow: extract a source URL into an evidence bundle, "
            "verify integrity, optionally notarize on-chain, then independently verify "
            "the signed action receipt."
        )
    )
    parser.add_argument(
        "--url",
        required=True,
        help="Source document URL to extract (PDF, HTML, etc.)",
    )
    parser.add_argument(
        "--notarize",
        action="store_true",
        help="Also notarize the bundle on-chain (uses plan quota / credits)",
    )
    parser.add_argument(
        "--allow-missing-receipt",
        action="store_true",
        help="Do not fail if the API returns no signed action receipt (dev only)",
    )
    args = parser.parse_args(argv)

    try:
        api_key, base_url = load_config()
        summary = run(
            url=args.url,
            notarize=args.notarize,
            api_key=api_key,
            base_url=base_url,
            allow_missing_receipt=args.allow_missing_receipt,
        )
    except DocImprintError as exc:
        print_error(exc)
        return 1
    except SystemExit as exc:
        return int(exc.code) if isinstance(exc.code, int) else 1

    json_dump({"ok": True, **summary})
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/typescript/src/production-flow.ts
/**
 * Evidence + receipt deep dive: URL extract → verify → optional notarize → receipt.
 *
 * Usage:
 *   npm run production-flow -- --url https://example.com/contract.pdf
 *   npm run production-flow -- --url https://example.com/contract.pdf --notarize
 */

import chalk from 'chalk'
import {
  apiVerifyUrl,
  bundleIdOf,
  createClient,
  extractUrl,
  loadConfig,
  manifestSha256,
  parseArgs,
  printError,
  printSharePanel,
  receiptIdOf,
  requireValidReceipt,
  verifyToolUrl,
  writeEvidencePack,
} from './shared.js'

async function main(): Promise<void> {
  const { notarize, allowMissingReceipt, url } = parseArgs(process.argv.slice(2))
  if (!url) {
    console.error(chalk.red('error:'), '--url is required')
    process.exit(1)
  }

  const { apiKey, baseUrl } = loadConfig()
  const client = createClient(apiKey, baseUrl)
  const warnings: string[] = []

  console.log(chalk.blue.bold('\nEvidence + receipt'))
  console.log('  Extract a URL into a signed bundle, verify integrity,')
  console.log('  then independently audit the action receipt.\n')

  const extracted = await extractUrl({ apiKey, baseUrl, url, retries: 3 })

  const bundleId = bundleIdOf(extracted)
  if (!bundleId) {
    throw new Error('extract response missing bundle_id')
  }
  const manifest = manifestSha256(extracted)
  console.log(`  bundle_id = ${chalk.cyan(bundleId)}`)
  console.log(`  manifest  = ${chalk.cyan(manifest ?? '—')}`)

  const verified = await client.verify(bundleId)
  if (!verified.valid) {
    throw new Error('bundle verification failed')
  }
  console.log(`  verify   = ${chalk.green('valid')}`)
  const inlineRid = receiptIdOf(verified.receipt ?? null)

  let notarizeSummary: Record<string, unknown> | null = null
  if (notarize) {
    const notarized = await client.notarize(bundleId)
    notarizeSummary = {
      tx_hash: notarized.tx_hash,
      network: notarized.network,
      eas_attestation_uid: notarized.eas_attestation_uid,
      attested_at: notarized.attested_at,
      receipt_id: receiptIdOf(notarized.receipt ?? null),
    }
    console.log('  notarize = done')
  }

  const { receiptVerify, receiptRows, warnings: receiptWarnings } = await requireValidReceipt(
    client,
    {
      bundleId,
      preferredId: inlineRid,
      allowMissing: allowMissingReceipt,
    },
  )
  warnings.push(...receiptWarnings)
  if (receiptVerify) {
    console.log(`  receipt  = ${chalk.green('valid')}`)
  }

  const summary = {
    demo: 'production_flow',
    source_url: url,
    bundle_id: bundleId,
    manifest_sha256: manifest ?? verified.manifest_sha256,
    verify: {
      valid: true,
      manifest_sha256: verified.manifest_sha256 ?? manifest,
      signature_valid: verified.signature_valid,
      artifacts_valid: verified.artifacts_valid,
    },
    notarize: notarizeSummary,
    inline_receipt_id: inlineRid ?? null,
    receipt_list_count: receiptRows.length,
    receipt_verify: receiptVerify,
    links: {
      verify_ui: verifyToolUrl(bundleId),
      verify_api: apiVerifyUrl(baseUrl, bundleId),
    },
    warnings,
  }

  const packPath = await writeEvidencePack(bundleId, summary)
  printSharePanel({
    bundleId,
    manifest: summary.manifest_sha256,
    baseUrl,
    receiptVerify,
    packPath,
  })
  for (const w of warnings) console.warn(chalk.yellow('warning:'), w)
  console.log(JSON.stringify({ ok: true, ...summary }, null, 2))
}

main().catch((err) => {
  printError(err)
  process.exit(1)
})
